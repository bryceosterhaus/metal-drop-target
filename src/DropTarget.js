import Component, {Config} from 'metal-jsx';
import Transition from 'metal-css-transitions';

export const DRAG_TYPES = {
	FILE: 'Files',
	HTML: 'text/html',
	TEXT: 'text/plain',
	URL: 'text/uri-list'
};

class DropTarget extends Component {
	created() {
		this.handleDocumentDrop = this.handleDocumentDrop.bind(this);
		this.handleDragEnd = this.handleDragEnd.bind(this);
		this.handleDragEnter = this.handleDragEnter.bind(this);
		this.handlePreventDefault = this.handlePreventDefault.bind(this);
		this.handleTargetDrop = this.handleTargetDrop.bind(this);
		this.handleTargetEnter = this.handleTargetEnter.bind(this);
		this.handleTargetLeave = this.handleTargetLeave.bind(this);
	}

	attached() {
		const {container} = this.props;

		container.addEventListener('dragend', this.handleDragEnd);
		container.addEventListener('dragenter', this.handleDragEnter);
		container.addEventListener('dragover', this.handlePreventDefault);
		container.addEventListener('drop', this.handleDocumentDrop);
		container.addEventListener('mouseleave', this.handleDragEnd);
	}

	detached() {
		const {container} = this.props;

		container.removeEventListener('dragend', this.handleDragEnd);
		container.removeEventListener('dragenter', this.handleDragEnter);
		container.removeEventListener('dragover', this.handlePreventDefault);
		container.removeEventListener('drop', this.handleDocumentDrop);
		container.removeEventListener('mouseleave', this.handleDragEnd);
	}

	handleDocumentDrop(event) {
		event.preventDefault();

		this.handleDragEnd();
	}

	handleDragEnd() {
		if (this.state.dragging) {
			this.setState({
				dragging: false,
				hoverOver: false
			});
		}
	}

	handleDragEnter(event) {
		const {types} = event.dataTransfer;

		if (types.indexOf(this.props.targetType) !== -1) {
			this.state.dragging = true;
		}
	}

	handleTargetDrop(event) {
		event.preventDefault();

		this.props.onDrop(event);

		this.handleDragEnd();
	}

	handleTargetEnter() {
		if (this.state.dragging) {
			this.state.hoverOver = true;
		}
	}

	handleTargetLeave() {
		if (this.state.dragging) {
			this.state.hoverOver = false;
		}
	}

	handlePreventDefault(event) {
		if (this.state.dragging) {
			event.preventDefault();
		}
	}

	render() {
		const {
			props: {children, dragMessage, hoverMessage},
			state: {dragging, hoverOver}
		} = this;

		return (
			<div
				class="drop-target-container"
				onDragEnter={this.handleTargetEnter}
				onDragLeave={this.handleTargetLeave}
				onDragOver={this.handlePreventDefault}
				onDrop={this.handleTargetDrop}
			>
				<Transition name="drop-zone-fade-in">
					{dragging && (
						<div class="drop-zone">
							{hoverOver ? hoverMessage : dragMessage}
						</div>
					)}
				</Transition>

				{children}
			</div>
		);
	}
}

DropTarget.PROPS = {
	container: Config.object().value(document),
	dragMessage: Config.any(),
	hoverMessage: Config.any(),
	onDrop: Config.func(),
	targetType: Config.oneOf(Object.values(DRAG_TYPES))
};

DropTarget.STATE = {
	dragging: Config.value(false),
	hoverOver: Config.value(false)
};

export default DropTarget;
