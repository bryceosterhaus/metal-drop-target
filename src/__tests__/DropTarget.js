import DropTarget, {DRAG_TYPES} from '../DropTarget';

const mockEventObj = {
	preventDefault() {}
};

describe('DropTarget', () => {
	let component;

	afterEach(() => {
		if (component) {
			component.dispose();
		}
	});

	it('renders', () => {
		component = new DropTarget();

		expect(component).toMatchSnapshot();
	});

	it('should call handleDragEnd', () => {
		component = new DropTarget();

		const spy = jest.fn();

		component.handleDragEnd = spy;

		component.handleDocumentDrop(mockEventObj);

		expect(spy).toBeCalled();
	});

	it('should reset state if currently dragging', () => {
		component = new DropTarget();

		component.state.dragging = true;
		component.state.hoverOver = true;

		component.handleDragEnd();

		expect(component.state.dragging).toBe(false);
		expect(component.state.hoverOver).toBe(false);
	});

	it('should set dragging to true if matches file type', () => {
		component = new DropTarget({
			targetType: DRAG_TYPES.FILE
		});

		component.handleDragEnter({
			dataTransfer: {
				types: ['text/html']
			}
		});

		expect(component.state.dragging).toBe(false);

		component.handleDragEnter({
			dataTransfer: {
				types: ['Files']
			}
		});

		expect(component.state.dragging).toBe(true);
	});

	it('should call onDrop and handleDragEnd', () => {
		const spy1 = jest.fn();
		const spy2 = jest.fn();

		component = new DropTarget({
			onDrop: spy1
		});

		component.handleDragEnd = spy2;
		component.state.dragging = true;

		component.handleTargetDrop(mockEventObj);

		expect(spy1).toBeCalled();
		expect(spy2).toBeCalled();
	});

	it('should set hoverOver to true if dragging', () => {
		component = new DropTarget();

		component.handleTargetEnter(mockEventObj);

		expect(component.state.hoverOver).toBe(false);

		component.state.dragging = true;

		component.handleTargetEnter(mockEventObj);

		expect(component.state.hoverOver).toBe(true);
	});

	it('should set hoverOver to false', () => {
		component = new DropTarget();

		component.state.dragging = true;
		component.state.hoverOver = true;

		component.handleTargetLeave();

		expect(component.state.hoverOver).toBe(false);
	});

	it('should call preventDefault if dragging', () => {
		const spy = jest.fn();

		component = new DropTarget();

		component.state.dragging = true;

		component.handlePreventDefault({
			preventDefault: spy
		});

		expect(spy).toBeCalled();
	});
});
