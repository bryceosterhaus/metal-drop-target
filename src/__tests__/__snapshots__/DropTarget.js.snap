// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropTarget renders 1`] = `
<div class="drop-target-container">
  <span></span>
</div>
`;
